### ISSUE TITLE
[🐞] Qwik `routeLoader$` loses data on resume

### ISSUE THREAD

**Issue by blecorre-mnp** [BuilderIO/qwik#4542]
### Which component is affected?

Qwik Runtime

### Describe the bug

Using Qwik in production websites for our company, we noticed something unusual. When getting for example user data from a `routeLoader$(...)` and storing it into a context, when the app is resumed on the browser, the context data seems to be deleted:

- All classes with any type of link to the context object's properties are deleted instantly.
- Displayed properties become blank.
- Only copied properties (`stringData.toUpperCase()`, `stringData.substring(...)`, etc) remain as is.

We are wondering if the error is on our side, or if there might be a problem with the implementation of `routeLoader$`.

*It is unfortunately very volatile*. It works on some machines, while on others it doesn't. The example below is a small test project. It doesn't work on my machine, but does in StackBlitz despite projects beeing a copy-paste of one another. It also only occurs when SSR is enabled, ie. running `npm run preview`, or in production environments. See screen capture below.

https://github.com/BuilderIO/qwik/assets/127991145/8850cf35-59cc-4a7f-94e2-93e50ecea35d

### Reproduction

https://stackblitz.com/edit/qwik-starter-eocsf7

### System Info

```shell
System:
    OS: macOS 13.3.1
    CPU: (8) arm64 Apple M2
    Memory: 108.41 MB / 8.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 19.8.1 - ~/.nvm/versions/node/v19.8.1/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 9.5.1 - ~/.nvm/versions/node/v19.8.1/bin/npm
  Browsers:
    Firefox: 113.0.1
    Safari: 16.4
  npmPackages:
    @builder.io/qwik: ^1.1.5 => 1.1.5 
    @builder.io/qwik-city: ^1.1.5 => 1.1.5 
    undici: 5.22.1 => 5.22.1 
    vite: 4.3.9 => 4.3.9
```

**Comment by blecorre-mnp** [comment]
Sorry if there's not that much info there, it is a pain to reproduce and investigate on our part...

**Comment by manucorporat** [comment]
When i navigate, it goes to:
http://localhost:5173/undefined/conversations

think i need some environment variable

**Comment by manucorporat** [comment]
Also, the app seems missing routes etc, can u please check again if the repo you provided is enough + some steps

**Comment by blecorre-mnp** [comment]
The environment variables are not useful in this case, it's only because the repro is a small part of the larger production project. Same for the routes, is most of them are redirecting to an older web platform.

**Comment by blecorre-mnp** [comment]
I'm sorry I really have nothing more to show you except that on some of our production environments running exactly the same build / serving commands, the issue occurs in the same way that right now on my machine.

**Comment by aroblast** [comment]
(from personnal account)

I found out what is going on, it is a bit more complicated than I thought. The data i'm fetching from my `routeLoader$` is at first available in the browser as expected. It is then destroyed when the browser resumes the application, eg:

```typescript
export const useGetData = routeLoader$(async () => {
    return fetchedData;
})

export default component$(() => {
    const data = useGetData();

    return <div>{data...}</div>
})
```

This code works fine on initial page load. Then, when the browser resumes the application, the data gets destroyed. Wrapping the fetched data as a `useSignal(...)` fixes the issue, but it is ugly as `routeLoader$()` is supposed to return a signal.

**I also noticed that this doesn't happen in a basic project**, but as complexity increases, it seems that at a certain point 
`routeLoader$` ceases to work correctly as a signal. It may be an issue somewhere in my code, but with no error logs on the console + debugger it is quite hard to know where..

@manucorporat does it ring a bell? Anyone feel free to get in touch if needed.

**Comment by blecorre-mnp** [comment]
Hey @OrenSayag,

I noticed yesterday evening while investigating at home that it only happens once I reach a certain level of complexity in my code base. I would love to give you access to the full repo, but it is company production code and I'm afraid I cant. I unfortunately don't have a lot of time right now to make a full scale reproduction of this issue, but I can maybe take some time later in the week to give you more details about this.

I have found a workaround, wrapping the `routeLoader$` result into a signal, and data is persisted again so it is not that urgent for us anymore. I don't know if anyone experienced the same issue with growing complexity?

Again it might well be my code doing something wrong, except that I can't really know with no logs and clear way to debug what Qwik is doing under the hood. I'll try to whip up a better reproduction this weekend for all of you. Thanks again though for your help and I hope to finally understand what's going on, kinda frustrating to not be able to use Qwik's full potential because of small bugs like this, although the workaround seems to work fine for now!

**Comment by blecorre-mnp** [comment]
Hey!

For everyone stumbling onto this issue in the future, I have finally found the reason why my `routeLoader$` is losing its data.

**tldr; using `async` methods in tasks like `routeAction$` or `globalAction$` might lead to loader's data getting erased.**

I was sometimes using async tasks and QRLs to update the state of the application. I don't really know how the `routeLoader$` works behind the scene, but once a `routeAction$` (maybe `globalAction$` too) is called, the data from the `routeLoader$` seems to be refreshed automatically. But in some cases, for example using async methods in `useVisibleTasks$` that update the loader's state in the backend using actions, something happens and the loader loses its data.

My fix behind this was to not use state updating actions in tasks, and find a workaround this problem. As an example, here is a small example of what not to do:

```
  useVisibleTask$(({ track }) => {
    track(() => travel.value);

    const travelCost = parseFloat(travel.value) as number;
    
    // Set travel cost here causes the underlying object storing this property in the backend
    // to get erased from its loader in the web app.
    if (travelCost >= 0) onSetTravelCost(travelCost); // <- async globalAction code behind.
  });
```

Defining the task callback as `async` breaks the whole task.

Hope this helps!
